### PR TITLE
Diet parser

### LINKED ISSUE #29: Add parser for diet

### PR DESCRIPTION
**Pull request by nihalzp**
Closes #29 and #30

### CODE REVIEW

File: src/main/java/athleticli/ui/Parser.java [hunk 1]
```diff
@@ -338,4 +353,178 @@ public static EditSleepCommand parseSleepEdit(String commandArgs) throws Athleti
...
 
+    /**
+     * Parses the raw user input for a diet and returns the corresponding diet object.
+     *
+     * @param commandArgs The raw user input containing the arguments.
+     * @return An object representing the diet.
+     * @throws AthletiException
+     */
+    public static Diet parseDiet(String commandArgs) throws AthletiException {
+        final String caloriesMarkerConstant = "calories/";
```
Comment (DaDevChia) [inline src/main/java/athleticli/ui/Parser.java:364]: Good use of String constants, but might be preferable to factor the String markers to a separate class to allow for easier changes to be made

File: src/main/java/athleticli/commands/diet/DeleteDietCommand.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+     * @param data The current data containing the diet list.
+     * @return The message which will be shown to the user.
+     */
+    public String[] execute(Data data) throws AthletiException {
+        if (index > data.getDiets().size() || index < 1) {
+            throw new AthletiException(Message.MESSAGE_INVALID_DIET_INDEX);
+        }
+        DietList dietList = data.getDiets();
+        int size = dietList.size();
+        Diet oldDiet = dietList.get(index - 1);
```
Comment (DaDevChia) [inline src/main/java/athleticli/commands/diet/DeleteDietCommand.java:37]: Would be preferable to rearrange this to the top of this method signature such that the exception checker can use index > size for better readability

File: docs/README.md
[the remark this answers is not in the document]
Comment (nihalzp) [reply docs/README.md:1]: Thanks for pointing it out. I have added a commit to resolve the issue.

File: src/main/java/athleticli/ui/Parser.java [hunk 3]
```diff
@@ -282,11 +295,12 @@ public static AddSleepCommand parseSleepAdd(String commandArgs) throws AthletiEx
...
-    
-        String startTime = commandArgs.substring(startMarkerPos + startMarkerConstant.length(), endMarkerPos).trim();
+
+        String startTime =
+                commandArgs.substring(startMarkerPos + startMarkerConstant.length(), endMarkerPos).trim();
         String endTime = commandArgs.substring(endMarkerPos + endMarkerConstant.length()).trim();
 
-        if(startTime.isEmpty() || endTime.isEmpty()) {
+        if (startTime.isEmpty() || endTime.isEmpty()) {
             throw new AthletiException("Please specify both the start and end time of your sleep.");
```
Comment (DaDevChia) [inline src/main/java/athleticli/ui/Parser.java:304]: Some AthletiExceptions can still be factored out as string constants in messages class

File: src/main/java/athleticli/ui/Parser.java [hunk 4]
```diff
@@ -338,4 +353,178 @@ public static EditSleepCommand parseSleepEdit(String commandArgs) throws Athleti
...
+    /**
+     * Checks if the user input for a diet contains all the required arguments.
+     *
+     * @param caloriesMarkerPos The position of the calories marker.
+     * @param proteinMarkerPos  The position of the protein marker.
+     * @param carbMarkerPos     The position of the carb marker.
+     * @param fatMarkerPos      The position of the fat marker.
+     * @throws AthletiException
+     */
+    private static void checkMissingDietArguments(int caloriesMarkerPos, int proteinMarkerPos,
```
Comment (DaDevChia) [inline src/main/java/athleticli/ui/Parser.java:404]: Great use of abstractions to perform check for missing arguments

File: src/test/java/athleticli/ui/ParserTest.java [hunk 5]
```diff
@@ -66,4 +67,106 @@ void parseCommand_listSleepCommand_expectListSleepCommand() throws AthletiExcept
         final String listSleepCommandString = "list-sleep";
         assertInstanceOf(ListSleepCommand.class, parseCommand(listSleepCommandString));
     }
+
+    @Test
+    void parseCommand_addDietCommand_expectAddDietCommand() throws AthletiException {
```
Comment (DaDevChia) [inline src/test/java/athleticli/ui/ParserTest.java:72]: Great check to test for exceptions thrown as well!

**Review by DaDevChia — APPROVED** [review]
Overall great code quality!

File: src/main/java/athleticli/ui/Parser.java [hunk 6]
```diff
@@ -338,4 +353,178 @@ public static EditSleepCommand parseSleepEdit(String commandArgs) throws Athleti
...
+    }
+
+    /**
+     * Parses the protein input for a diet.
+     *
+     * @param protein The protein input.
+     * @return The parsed protein.
+     * @throws AthletiException
+     */
+    public static int parseProtein(String protein) throws AthletiException {
```
[the remark this answers is not in the document]
Comment (nihalzp) [reply src/main/java/athleticli/ui/Parser.java:470]: True. However, the current way is convenient to send personalised error messages.

File: src/main/java/athleticli/ui/Parser.java (hunk 1 above)
Comment (nihalzp) [reply src/main/java/athleticli/ui/Parser.java:364]: Good idea. We can open an issue for it.

File: src/main/java/athleticli/commands/diet/DeleteDietCommand.java (hunk 2 above)
Comment (nihalzp) [reply src/main/java/athleticli/commands/diet/DeleteDietCommand.java]: Good idea. Resolved.

**Pull request merged**